### PR TITLE
python37Packages.fastapi: 0.45.0 -> 0.49.0

### PR DESCRIPTION
**Pull request by wd15**
###### Motivation for this change

 - Pin Starlette to verion 0.12.9 due to tiangolo/fastapi#683 otherwise fails to build.

 - Add peewee to test environment as now required.

 - Remove pyproject.toml patching as no longer required.

First commit is part of #81304 so that needs to be merged first

These changes need to backported to release-20.03, #80379

This PR partially replaces #81055

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by wd15** [comment]
Result of `nix-review pr 81305` [1](https://github.com/Mic92/nix-review)
<details>
  <summary>2 package were build:</summary>

  - python37Packages.fastapi
  - python37Packages.starlette
</details>

File: pkgs/development/python-modules/fastapi/default.nix [hunk 1]
```diff
@@ -10,24 +11,34 @@
...
   propagatedBuildInputs = [
     uvicorn
-    starlette
+    (starlette.overridePythonAttrs (old: rec {
+      version = "0.12.9";
+      src = fetchurl {
+        url = "https://github.com/encode/starlette/archive/${version}.tar.gz";
+        sha256 = "0d3g44hajqmjlr46mf527algqf6a5zh7k7dssl6hwzqm71qr09jp";
+      };
+    }))
```
Comment (jonringer) [inline pkgs/development/python-modules/fastapi/default.nix]: not okay. this will cause issues if another package brings in starlette at latest. You should patch instead you want to patch the version.

You should be able to be within a major version, and not be broken (on packages that adhere to semver)

example:
https://github.com/NixOS/nixpkgs/blob/24720dd250cb26025d6dd5dbaf089d48f2493cb5/pkgs/development/python-modules/celery/default.nix#L14

File: pkgs/development/python-modules/fastapi/default.nix [hunk 2]
```diff
@@ -37,16 +48,9 @@ buildPythonPackage rec {
...
     aiosqlite
+    peewee
   ];
 
-  # starlette pinning kept in place due to 0.12.9 being a hard
-  # dependency luckily fastapi is currently the only dependent on
-  # starlette. Please remove pinning when possible
-  postPatch = ''
-    substituteInPlace pyproject.toml \
-      --replace "pydantic >=0.32.2,<=0.32.2" "pydantic"
```
Comment (jonringer) [inline pkgs/development/python-modules/fastapi/default.nix]: please do this version patching for starlette

File: pkgs/development/python-modules/fastapi/default.nix (hunk 1 above)
Comment (wd15) [reply pkgs/development/python-modules/fastapi/default.nix]: Thanks for reviewing. Unfortunately, Fastapi is [completely broken](https://github.com/tiangolo/fastapi/issues/683#issuecomment-585931388) for starlette > 0.12.13. Thus, using the latest version of Starlette with Fastapi. I have three choices,

 - revert the version of Starlette in Starlette's own default.nix to be something compatible (Fastapi is Starlette's only dependent)
 - leave this broken for now until fastapi works correctly with later versions of Starlette
 - remove fastapi from Nixpkgs until fixed properly

I think I'll go with the first one unless there's an objection.

**Comment by wd15** [comment]
I've reverted to fastapi version 0.12.9.

**Comment by jonringer** [comment]
@GrahamcOfBorg eval

File: pkgs/development/python-modules/starlette/default.nix [hunk 3]
```diff
@@ -20,12 +20,18 @@
...
+  # possible. FastAPI is currently Starlette's only dependent.
+
+  version = "0.12.9";
   disabled = isPy27;
 
-  src = fetchPypi {
-    inherit pname version;
-    sha256 = "6bd414152d40d000ccbf6aa40ed89718b40868366a0f69fb83034f416303acef";
+  src = fetchurl {
+    url = "https://github.com/encode/starlette/archive/${version}.tar.gz";
```
Comment (jonringer) [inline pkgs/development/python-modules/starlette/default.nix]: please use fetchFromGitHub helper
```suggestion
  src = fetchFromGitHub {
    owner = "encode";
    repo = pname;
    rev = version;
```

otherwise LGTM

Comment (wd15) [reply pkgs/development/python-modules/starlette/default.nix]: Done now. Should be ready to merge.

File: pkgs/development/python-modules/starlette/default.nix [hunk 4]
```diff
@@ -20,12 +20,20 @@
...
+  # https://github.com/tiangolo/fastapi/issues/683. Please update when
+  # possible. FastAPI is currently Starlette's only dependent.
+
+  version = "0.12.9";
   disabled = isPy27;
 
-  src = fetchPypi {
-    inherit pname version;
-    sha256 = "6bd414152d40d000ccbf6aa40ed89718b40868366a0f69fb83034f416303acef";
+  src = fetchFromGitHub {
```
Comment (rnhmjoj) [inline pkgs/development/python-modules/starlette/default.nix]: Why is these needed?

Comment (wd15) [reply pkgs/development/python-modules/starlette/default.nix]: There was some discussion about [this here](https://github.com/NixOS/nixpkgs/pull/48633#issuecomment-432732045) and other links that I can't find now. A standard Python distribution doesn't have the test suite as part of the distribution. Ideally, Python distributions would have some standard for self testing, but they don't.

Comment (rnhmjoj) [reply pkgs/development/python-modules/starlette/default.nix]: Ah, ok for the tests. Thank you.

File: pkgs/development/python-modules/fastapi/default.nix (hunk 1 above)
Comment (jonringer) [reply pkgs/development/python-modules/fastapi/default.nix]: reverting is fine, just unfortunate. I think this is reasonable.

**Review by jonringer — APPROVED** [review]
diff LGTM
commit LGTM

```
[4 built, 15 copied (4.7 MiB), 1.2 MiB DL]
https://github.com/NixOS/nixpkgs/pull/81305
2 package built:
python37Packages.fastapi python37Packages.starlette
